### PR TITLE
Add Linear Nk OCCRI method + ISDF

### PR DESCRIPTION
**Pull request by kosm6966**
##Summary

  • Add OCCRI method for efficient exact exchange evaluation in periodic systems with significant algorithmic improvements. 
        • Performance is similar to but typically faster than existing occri method
        • Designed to integrate with isdfx
        • Provide memory-efficient implementations with MO blocking following PySCF patterns
  • Implement ISDFX (Interpolative Separable Density Fitting eXchange) module with comprehensive testing and examples
        • Exchange build is linear with N_k as shown by Example 2
        • Default threshold is set and provides < 50 μHa error for all test systems
  • Include comprehensive test suites and educational examples demonstrating usage and performance

##  Features

  • OCCRI: Exchange matrix evaluation using occupied orbital resolution of identity
  • ISDFX: Tensor hypercontraction-based exchange with interpolative separable density fitting
  • Support for RHF, UHF, RKS, UKS and k-point variants (KRHF, KUHF, KUKS, KRKS)
  • Memory management with automatic MO blocking for large systems
  • Comprehensive utility functions separated from object methods for better code organization
  • Natural orbital construction and exchange matrix building utilities

##  Performance & Accuracy

  • ISDFX provides near-FFTDF accuracy (typically <50 μHa energy differences)
  • Scaling analysis shows sub-quadratic performance with k-point sampling
  • Memory-efficient MO blocking prevents memory overflow for large systems
  • Proper handling of asymmetric k-point meshes and complex conjugation

##  Testing & Examples

  • 84 comprehensive tests covering energy accuracy, k-point handling, and edge cases
  • Energy comparison tests with 50 μHa tolerance vs FFTDF reference
  • Asymmetric k-point tests to catch symmetry and conjugation bugs
  • Educational examples: basic usage, AFM initial guess setup, scaling analysis
  • Integration with existing run_tests.py framework with --perf flag support

##  Code Quality

  • Follows PySCF coding conventions with comprehensive docstrings
  • Separation of pure algorithms from object methods for better maintainability
  • Formatted with Black, isort, and flake8
  • Extensive error handling and input validation
  • Memory cleanup following PySCF patterns

##  Test plan

  - All 84 tests pass including energy accuracy and performance tests
  - Examples demonstrate proper usage, AFM setup, and scaling behavior
  - Code quality checks pass (Black, isort, flake8)
  - Memory management validated with MO blocking implementation
  - K-point handling thoroughly tested with asymmetric meshes

### CODE REVIEW

**Comment by jeanwsr** [comment]
Is it better to put these codes in pbc/df?

File: pyscf/occri/test/run_tests.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (kosm6966) [reply pyscf/occri/test/run_tests.py:1]: removed in 532908e

File: pyscf/occri/__init__.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        self.method = mydf.__module__.rsplit(".", 1)[-1]
+        assert self.method in ["hf", "uhf", "khf", "kuhf", "rks", "uks", "krks", "kuks"]
+
+        self.StartTime = time.time()
+        self.cell = mydf.cell
+        kpts = mydf.kpts
+        if kpts is None:
+            kpts = numpy.zeros(3, numpy.float64)
+        kpts = kpts.round(6)
+        self.kmesh = [
```
[the remark this answers is not in the document]
Comment (kosm6966) [reply pyscf/occri/__init__.py]: using kpts_to_kmesh from k2gamma [512122f].

File: pyscf/lib/occri/occri.c [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+     * mesh : int[3]
+     *     FFT mesh dimensions [nx, ny, nz]
+     * nmo, nao, ngrids : int
+     *     Number of MOs, AOs, and grid points
+     */
+    
+    const int nthreads = omp_get_max_threads();
+
+    // Initialize FFTW multithreading support
+    if (!fftw_init_threads()) {
```
[the remark this answers is not in the document]
Comment (kosm6966) [reply pyscf/lib/occri/occri.c]: removed [532908e].

**Review by sunqm — APPROVED** [review]
